compiler-core/templates/prelude.js

```diff
@@ -116,6 +116,9 @@ export class BitString {
   }
 
   constructor(buffer) {
+    if (!(buffer instanceof Uint8Array)) {
```

Comment:
Shouldn't this be `ArrayBuffer`? As per https://github.com/gleam-lang/gleam/pull/2290/files#r1283405089?

As that's the underlying storage and Uint8Array is a wrapper.

Reply:
That was the original plan, but everywhere you construct one (for instance in the prelude tests), you pass in a `Uint8Array`. I have to imagine folks writing ffi code right now are doing the same.
I think either is a reasonable choice, but this seemed like it would break fewer uses. Going to ask in the discord and see if anyone other than us is currently manually constructing `BitString`s in js

Reply:
I suspect very few people are constructing bit strings directly, and now is the time to make a breaking change.

Having said that, couldn't we do something like `thing.buffer` for any of the typed array classes?

Reply:
Alright, I think I've come full circle on this, and it's served as a lesson to try a few different things before considering making a pull request haha. 
Lets use just `ArrayBuffer` and throw an exception if it's anything else. 
`ArrayBuffer` is the actual chunk of memory, the simplest piece, so I think it's what should underpin the internals (as you stated before). 
Encouraging an API that takes any Typed Array like a `Uint8Array` or a `DataView` and implicitly pulling out the buffer is probably a bad idea, because the view might be only of segment of the underlying buffer.
I'm going to update the `BitString` class to be based around an `ArrayBuffer` and push for your review.

Reply:
Ah right! That makes much more sense, thank you.

>  because the view might be only of segment of the underlying buffer.

This is the key bit here- we need to be able to support having bit strings being a slice in to the underlying buffer, to avoid having lots of copying when iterating over a bit string or slicing new bit strings from them. I foolishly didn't understand the relationship between these classes and slices.

With that constraint in mind what's the best one to use? If we used `ArrayBuffer` we would need to implement the logic ourselves for only using a portion of it, which could be slower than an optimised implementation that comes from one of the other classes.

Reply:
Right now I have an ArrayBuffer backed impl. For the various operations it uses Uint8Array internally to work with the buffer, but the actual stored buffer is an ArrayBuffer. 
I edited unequalBuffer so the isEqual function works for any ArrayBuffer backed item, and now the only issue is that inlining a BitString into another is failing so I need to fix that.

Reply:
I honest to god think we could merge this as is. It's not like we ran into some actual performance constraint, I just didn't understand I needed to use a Uint8Array, but now the constructor makes that much more explicit. I think limiting to Uint8Array makes perfect sense, as it's the closest thing JavaScript has to a BitString, meets your requirements of cheaply creating multiple BitStrings from the same memory, and seems to be how the class was designed in the first place. I say we take the small win, be grateful for having the chance to learn more about JavaScript together, and focus energy on more important tasks lol.